contracts/test/EscrowEchidnaTest.sol

```diff
@@ -0,0 +1,101 @@
+        escrow = new L1Escrow();
+        escrow.approve(address(dai), address(this), type(uint256).max);
+        alice = address(new Alice(address(dai), address(this)));
+    }
+
+    // --- Math ---
+    function add(uint256 x, uint256 y) internal pure returns (uint256 z) {
+        z = x + y;
+        assert (z >= x); // check for addition overflow
+    }
+    function sub(uint256 x, uint256 y) internal pure returns (uint256 z) {
+        z = x - y;
+        assert (z <= x); // check for subtraction underflow
+    }
+
+    /// @dev Test that supply and balance hold on mint
+    function mint(uint256 wad) public {
+        uint256 supply = dai.totalSupply();
+        uint256 aliceBalance = dai.balanceOf(alice);
+        wad = 1 + wad % sub(MAX_SUPPLY, supply);
+        dai.mint(alice, wad);
+        assert(dai.balanceOf(alice) == add(aliceBalance, wad));
+        assert(dai.totalSupply() == add(supply, wad));
+    }
+
+    /// @dev Test that supply and balance hold on burn
+    function burn(uint256 wad) public {
+        uint256 supply = dai.totalSupply();
+        uint256 aliceBalance = dai.balanceOf(alice);
+        wad = aliceBalance == 0 ? 0 : 1 + wad % aliceBalance;
+        dai.burn(alice, wad);
+        assert(dai.balanceOf(alice) == sub(aliceBalance, wad));
+        assert(dai.totalSupply() == sub(supply, wad));
+    }
+
+    /// @dev Test that supply and balance hold on move
+    function move(uint256 wad) public {
+        uint256 aliceBalance = dai.balanceOf(alice);
+        uint256 bobBalance = dai.balanceOf(bob);
+        wad = aliceBalance == 0 ? 0 : 1 + wad % aliceBalance;
+        dai.move(alice, bob, wad);
+        assert(dai.balanceOf(alice) == sub(aliceBalance, wad));
+        assert(dai.balanceOf(bob) == add(bobBalance, wad));
+    }
+
+    /// @dev Test that supply and balance hold on deposit
+    function deposit(uint256 wad) public {
+        uint256 aliceBalance = dai.balanceOf(alice);
+        uint256 escrowBalance = dai.balanceOf(address(escrow));
+        wad = aliceBalance == 0 ? 0 : 1 + wad % aliceBalance;
+        dai.transferFrom(alice, address(escrow), wad);
+        oDai.mint(alice, wad);
+        assert(dai.balanceOf(alice) == sub(aliceBalance, wad));
+        assert(dai.balanceOf(address(escrow)) == add(escrowBalance, wad));
+        assert(dai.balanceOf(address(escrow)) == oDai.totalSupply());
+    }
+
+    /// @dev Test that supply and balance hold on withdraw
+    function withdraw(uint256 wad) public {
+        uint256 aliceBalance = oDai.balanceOf(alice);
+        uint256 escrowBalance = dai.balanceOf(address(escrow));
+        wad = aliceBalance == 0 ? 0 : 1 + wad % aliceBalance;
+        oDai.burn(alice, wad);
+        dai.transferFrom(address(escrow), alice, wad);
+        assert(dai.balanceOf(address(escrow)) == sub(escrowBalance, wad));
+        assert(oDai.balanceOf(alice) == sub(aliceBalance, wad));
+        assert(dai.balanceOf(address(escrow)) == oDai.totalSupply());
+    }
```

Comment:
My only comment here is that most of this code isn't doing anything meaningful--all you're actually checking is `mint`/`burn`/`transferFrom` on the `oDai` contract. The stuff with the escrow is just obfuscation of that since you aren't actually using the bridge functionality--you're just fuzzing your own test code at that point. It still provides coverage, so it doesn't really need to be changed, but it is more complicated than it needs to be. It's also potentially misleading--you should probably rename this to `ODaiEchidnaTest`.

Reply:
There is already the DaiEchidnaTest for L2 Dai.
The intention behind the EscrowEchidnaTest was to simulate real case deposits and withdraws to overcome limitations with echidna and complexity of the codebase to fuzz.
The approach uses the Escrow contract and the two Dai contract to fuzz deposits and withdraws in real case scenarios by allowing Alice and Bob to interact on L1 and Test that supply and balance hold on deposits and withdrawals.

Reply:
But you exercise none--absolutely none--of the actual deposit/withdraw logic of the bridges. Instead you manually transfer to and from the Escrow. This whole PR is a just a self-contained circular exercise.The Escrow could be any address that has approved the test contract to transfer its DAI, and the whole thing would still work.

Reply:
The EscrowEchidnaTest original intention was to fuzz the deposit/withdraw logic of the bridges (prev. gates) by using a MockMessenger to allow interactions between the two bridges. After trial and error it seems not to reach the bridge2 `finalizeDeposit` method to complete the deposit on L2. The current version propose a mock of deposit and withdraw methods in the tests to show properties and invariants of the Escrow and L2 Dai contracts.